### ISSUE TITLE
json templates for MQTT component will, birth, shutdown and prefix

### ISSUE THREAD

**Issue by Bazmundi** [esphome/feature-requests#1141]
**Describe the problem you have/What new integration you would like**

In a similar fashion to mqtt.publish_json, I would like to be able to build a json based will, birth, shutdown and perhaps even topic_prefix

**Please describe your use case for this integration and alternatives you've tried:**

The textual payloads offered for will, birth and shutdown do not allow parameterised json as a response, that might be achieved using a lambda.  The may be the occasional call for templated topic_prefix.

**Additional context**

**Comment by hansonap** [comment]
I agree that this would be useful, for instance adding the mac address to the payload.
